daemon/containerd/soft_delete.go

```diff
@@ -0,0 +1,72 @@
+package containerd
+
+import (
+	"context"
+
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/docker/docker/errdefs"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+// softImageDelete deletes the image, making sure that there are other images
+// that reference the content of the deleted image.
+// If no other image exists, a dangling one is created.
+func (i *ImageService) softImageDelete(img containerdimages.Image) error {
+	// Don't accept ctx as argument as we don't want this
+	// to be interrupted in the middle.
+	ctx := context.Background()
+	is := i.client.ImageService()
+
+	logger := logrus.WithFields(logrus.Fields{
+		"name":   img.Name,
+		"digest": img.Target.Digest.String(),
+	})
+
+	// If the image already exists, persist it as dangling image
+	// but only if no other image has the same target.
+	digest := img.Target.Digest.String()
+	imgs, err := is.List(ctx, "target.digest=="+digest)
+	if err != nil {
+		logger.WithField("digest", digest).WithError(err).Debug("failed to check if there are images targeting this digest")
+		return errdefs.System(err)
+	}
+
+	// Create dangling image if this is the last image pointing to this target.
+	if len(imgs) == 1 {
+		danglingImage := img
+
+		danglingImage.Name = danglingImageName(img.Target.Digest)
+		delete(danglingImage.Labels, "io.containerd.image.name")
+		delete(danglingImage.Labels, "org.opencontainers.image.ref.name")
+
+		_, err = is.Create(ctx, danglingImage)
+
+		// Error out in case we couldn't persist the old image.
+		// If it already exists, then just continue.
+		if err != nil && !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Error("failed to create a dangling image for the image being replaced")
+			return errdefs.System(err)
+		}
+	}
+
+	// Free the target name.
+	err = is.Delete(ctx, img.Name)
+	if err != nil {
+		if !cerrdefs.IsNotFound(err) {
+			logger.WithError(err).Error("failed to delete image which existed a moment before")
+			return errdefs.System(err)
+		}
+	}
+
+	return nil
+}
+
+func isDanglingImage(image containerdimages.Image) bool {
+	return image.Name == danglingImageName(image.Target.Digest)
+}
```

Comment:
Wondering if `hasPrefix("dangling@")` would be enough for these 

edit: actually; it looks like `isDanglingImage()` is not used anywhere; was it meant to be used anywhere?

Reply:
It would be enough, but it would also match images which are unlikely to be created by us. User could fiddle with the containerd namespace him self and create an image with name `dangling@home` - which isn't a valid reference, and is not a "dangling" image which we understand.
The whole "dangling" image is our convention, so IMO it's better no to be too liberal in what we consider to be a dangling image.

Yeah it was used first for this branch check: https://github.com/moby/moby/blob/eaf7acbc2951dc40960ceb2e6a436a4494085a07/daemon/containerd/image_tag.go#L69-L80

but then I changed it and it's not used, but it since this is the first upstream PR that creates dangling images I kept it because it will be used by other things (could be used by https://github.com/moby/moby/pull/44840#discussion_r1084202281)